selfdrive/test/process_replay/test_processes.py

```diff
@@ -154,38 +182,41 @@ def format_diff(results, ref_commit):
     untested = (set(interface_names) - set(excluded_interfaces)) - tested_cars
     assert len(untested) == 0, f"Cars missing routes: {str(untested)}"
 
-  results: Any = {}
+  pool = NestablePool(args.jobs)
+
+  lreaders: Any = {}
+  p1 = pool.map_async(get_logreader, [seg for car, seg in segments])
+  for tup in p1.get():
+    if tup[0] not in lreaders:
+      lreaders[tup[0]] = {}
+    lreaders[tup[0]] = tup[1]
+
+  pool_args: Any = []
   for car_brand, segment in segments:
     if (len(args.whitelist_cars) and car_brand.upper() not in args.whitelist_cars) or \
        (not len(args.whitelist_cars) and car_brand.upper() in args.blacklist_cars):
       continue
-
-    print(f"***** testing route segment {segment} *****\n")
-
-    results[segment] = {}
-
-    r, n = segment.rsplit("--", 1)
-    lr = LogReader(get_url(r, n))
-
     for cfg in CONFIGS:
       if (len(args.whitelist_procs) and cfg.proc_name not in args.whitelist_procs) or \
          (not len(args.whitelist_procs) and cfg.proc_name in args.blacklist_procs):
         continue
-
       cur_log_fn = os.path.join(FAKEDATA, f"{segment}_{cfg.proc_name}_{cur_commit}.bz2")
-      if not args.upload_only:
-        ref_log_fn = os.path.join(FAKEDATA, f"{segment}_{cfg.proc_name}_{ref_commit}.bz2")
-        results[segment][cfg.proc_name], log_msgs = test_process(cfg, lr, ref_log_fn, args.ignore_fields, args.ignore_msgs)
-
-        # save logs so we can upload when updating refs
-        save_log(cur_log_fn, log_msgs)
-
+      pool_args.append((segment, cfg, args, cur_log_fn, lreaders[segment]))
       if upload:
         print(f'Uploading: {os.path.basename(cur_log_fn)}')
         assert os.path.exists(cur_log_fn), f"Cannot find log to upload: {cur_log_fn}"
         upload_file(cur_log_fn, os.path.basename(cur_log_fn))
         os.remove(cur_log_fn)
 
+  results: Any = {}
+  p2 = pool.map_async(run_test_process, pool_args)
+  for tup in p2.get():
+    if tup[0] not in results:
+      results[tup[0]] = {}
+    results[tup[0]][tup[1]] = tup[2]
+  pool.close()
+  pool.join()
```

Comment:
have you tried to wrap these in unittest/pytest test cases instead? I think this would have a few benefits:
* nice setup/teardown hooks
* cleaner stdout. everything's jumbled together now
* parallel running for free with pytest-xdist

Reply:
I will look into this on another branch